internal/sync/lifecycle.go

```diff
@@ -56,13 +56,13 @@ const (
 // LifecycleOnce is a helper for implementing transport.Lifecycles
 // with similar behavior.
 type LifecycleOnce struct {
-	lock sync.Mutex
-
-	// TODO(abg): We don't need a mutex if we use atomics correctly. Right
-	// now, the only reason we have an atomic is so that we can safely query
-	// the state of the lifecycle while Start() is still running. We should
-	// just switch this to an atomic instead.
-
+	// We use the sync.Mutex in order to guarantee that multiple calls
+	// to Start/Stop will wait until the first call finishes and return
+	// the same error (stored in startErr/stopErr).
+	//
+	// State is stored in an atomic so that we can read it from the `IsRunning`
+	// function without having to worry about race conditions with Start/Stop.
```

Comment:
Also mention: "A RWMutex is not used because we don't want IsRunning to wait until Start/Stop are finished."

FWIW you *could* use an RWMutex; you would just have to have to separate locks: one mutex to make sure only one of start/stop was running at a time, and another for when you actually read/write state.

So like,

```
Start() {
    l.lock.Lock()
    defer l.lock.Unlock()
    // ...
    l.stateLock.Lock()
    l.state = ...
    l.stateLock.Unlock()
}

IsRunning() {
    l.stateLock.RLock()
    state := l.state
    l.stateLock.RUnlock()
    return state == ..
}
```